src/overlays/actors/ovl_Dm_Hina/z_dm_hina.c

```diff
@@ -33,24 +45,164 @@ const ActorInit Dm_Hina_InitVars = {
+    s32 i;
+    s16 light1Color;
+
+    switch (this->csState) {
+        case LIGHT_ORB_STATE_OFF:
+            this->lightOrbAlpha = 0;
+            this->scale = 0.0f;
+            break;
+
+        case LIGHT_ORB_STATE_ON:
+            Math_SmoothStepToF(&this->scale, 1.0f, 0.4f, 0.05f, 0.001f);
+            this->lightOrbAlpha = this->scale * 255.0f;
+
+            this->drawOrbEffect = Math_SinS(play->state.frames * 8000);
+
+            for (i = 0; i < 3; i++) {
+                light1Color = this->scale * -255.0f;
+                play->envCtx.lightSettings.diffuseColor1[i] = light1Color;
+                play->envCtx.lightSettings.fogColor[i] = light1Color;
+                play->envCtx.lightSettings.ambientColor[i] = light1Color;
+            }
+
+            play->envCtx.lightSettings.fogNear = this->scale * -500.0f;
+
+            if (play->envCtx.lightSettings.fogNear < -300) {
+                play->roomCtx.curRoom.segment = NULL;
+            }
+
+            break;
+    }
+}
+
+void DmHina_Update(Actor* thisx, PlayState* play) {
+    DmHina* this = THIS;
+
+    this->actionFunc(this, play);
+    DmHina_UpdateLightOrb(this, play);
+}
+
+void DmHina_DrawLightOrb(DmHina* this, PlayState* play) {
+    GraphicsContext* gfxCtx = play->state.gfxCtx;
+
+    if (this->csState != LIGHT_ORB_STATE_OFF) {
+        OPEN_DISPS(gfxCtx);
+
+        func_8012C2DC(play->state.gfxCtx);
+        gDPSetPrimColor(POLY_XLU_DISP++, 0, 0, 255, 255, (u8)(this->scale * 100.0f) + 155, this->lightOrbAlpha);
+        gDPSetEnvColor(POLY_XLU_DISP++, 255, 255, (u8)(this->drawOrbEffect * 100.0f) + 50, 0);
+        gSPDisplayList(POLY_XLU_DISP++, gLightOrb1DL);
+        Matrix_Translate(this->actor.world.pos.x,
+                         this->actor.world.pos.y + (this->oscilationFactor * this->yDimScaleFactor) +
+                             (40.0f * this->yDimScaleFactor),
+                         this->actor.world.pos.z, MTXMODE_NEW);
+        Matrix_ReplaceRotation(&play->billboardMtxF);
+        Matrix_Scale(this->scale * 20.0f, this->scale * 20.0f, this->scale * 20.0f, MTXMODE_APPLY);
+        Matrix_RotateZF(Rand_ZeroFloat(2 * M_PI), MTXMODE_APPLY);
+        gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+        gSPDisplayList(POLY_XLU_DISP++, gLightOrbVtxDL);
+
+        CLOSE_DISPS(gfxCtx);
+    }
+}
+
+void DmHina_Draw(Actor* thisx, PlayState* play) {
+    DmHina* this = THIS;
+    f32 scale;
+
+    if (this->isVisible) {
+        Matrix_Translate(this->actor.world.pos.x,
+                         this->actor.world.pos.y + (this->oscilationFactor * this->yDimScaleFactor) +
+                             (40.0f * this->yDimScaleFactor),
+                         this->actor.world.pos.z, MTXMODE_NEW);
+        Matrix_RotateZYX(0, play->gameplayFrames * 1000, 0, MTXMODE_APPLY);
+        scale = this->maskScale * (1.0f - this->scale) * this->yDimScaleFactor;
```

Comment:
I'm not convinced by the names of any of the variables on the right-hand-side of this. Since the actual mask's scale is proportional to `1 - this->scale`, the latter seems incorrectly named (seems like it's more like the progress of the light orb scale towards 0 from its original 20, so as the light orb grows this shrinks). The `yDimScaleFactor`'s primary use seems to be as the amplitude of the vertical oscillation, although it also features in other places, the whole calculation is a bit peculiar.